ASoC: SOF: ipc4-topology: Use the single format specified by the dai copier if possible

ujfalusi opened the pull request:
Hi,

needs comment and commit message updates, just to show the idea I had regarding to #4867 

I think there is one issue that this might have is if the DAI copier for DMIC have multiple bit depths then it is not going to force the 32bit, but forcing unconditionally 32bit blob w/o knowing if the copier have support for it in the first place is not correct.
what if we have 32bit blob and the copier only supports 16bit bit depth?

I think if we load the 32bit blob forcibly then we need to change the ref_params so that we will be looking for 32bit cormat support on the dai copier...

@ranj063, any thoughts?

ujfalusi committed:
ASoC: SOF: ipc4-topology: Correct DAI copier config and NHLT blob request

In case of capture and when the DAI copier have single bit depth supported
on it's input side we should use this format instead of the one in
fe_params.

Regardless of the stream direction for the NHLT blob lookup when the DAI
copier only supports single bit depth on the DAI side we should only look
for a blob which matches with this single configuration.

For DMIC if the DAI copier supports multiple bit depths, try to request
32-bit blob first if the requested bit depth is 16-bit.
If the 32-bit blob is available then look for marching (32-bit) copier
format to make sure that both the blob and copier have correct parameters.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Allow selective update in sof_ipc4_update_hw_params

Add a bitmask parameter to sof_ipc4_update_hw_params() to be able to select
the param to be updated.
This feature can be used when not all params should be updated, for example
if caller only wants to update the format in the params, leaving the
channels and rates untouched.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- fix i386 build ( sof_ipc4_prepare_copier_module() is a monster function... )

plbossart commented:
SOFCI TEST

ujfalusi commented:
Changes since v1:
- Fix bug found by @singalsu : the sof_ipc4_update_hw_params() should be called with **BIT(**SNDRV_PCM_HW_PARAM_FORMAT**)** when fixing up the format of the params in case of single format
- Tested by @singalsu 

TODO:
commit message and comment updates
Fix corner case when we want to pick 32bit format on DAI side but the copier declares multiple bit depths (atm we would go with the fe_params format).

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1490,7 +1477,8 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct sof_ipc4_copier_data *copier_data;
-	struct snd_pcm_hw_params ref_params;
+	struct snd_pcm_hw_params *ref_params;
```
@ujfalusi this is so confusing. Why do you change this to a local copy in the first patch only to change it in the follow up?

ujfalusi replied:
@ranj063, yes, I know, I was taking a different direction then turned back. It is gone in the new version.

singalsu reviewed (approved):
I re-checked this version, works OK. Commit text of 1st is not explaining all changes included?

ujfalusi commented:
Changes since v2:
- reduce the patch count to two
- Move the dai_copier preparation code out to a separate function
- The NHLT blob lookup now updates the params it receives in case different format is used to pick the blob (to find matching input settings)
- Comments updated, added
- Commit message proposal added
- Tested by @singalsu

ujfalusi commented:
> I re-checked this version, works OK. Commit text of 1st is not explaining all changes included?

Let me check the commit message, thanks for testing it!

plbossart commented:
I can't figure out what the second commit is "replace, move the original here! "ASoC: SOF: ipc4-topology: Look for …
…the 32-bit DMIC blob first"" ??

nor if it's better than https://github.com/thesofproject/linux/pull/4867

Shouldn't there be some discussion on alternatives and what the best option is?

ujfalusi commented:
> I can't figure out what the second commit is "replace, move the original here! "ASoC: SOF: ipc4-topology: Look for … …the 32-bit DMIC blob first"" ??

That is a replacement patch for the one which requested 32-bit blob for DMIC. For upstream the two patch needs to be combined with an updated commit message as a new commit.
Yeah, I could have just reverted the other in this PR...

> nor if it's better than #4867

This does things in revers and it is not built on assumptions. You see: with #4867 we would unconditionally select 32bit blob for DMIC even if the copier does not support 32bit. I know that it is, but that is an assumption, right?
This PR does things in reverse: it looks for a blob which matches with the single format supported by the copier, so indirectly we will have 32bit blob if the copier only supports 32bit, but this PR also handles the case when the copier have multiple formats, we will look for 32bit blob and if it is there we will look for the 32bit copier config and not a probably mismatching one from fe_params

> Shouldn't there be some discussion on alternatives and what the best option is?

Certainly!

ujfalusi commented:
Complete fail on MTLP_SDW_AIOC?

ujfalusi commented:
SOFCI TEST

ujfalusi commented:
USB network is not probing on MTL?

ujfalusi commented:
On a local laptop:
On boot the usb dongle is not detected (usb-c -> usb-a -> gigabit).
unplug / plug -> nothing in dmesg
lsusb shows other devices
Plug the dongle to other C port and it got detected and now in the original port it is working after boot.

ujfalusi committed:
Revert "ASoC: SOF: ipc4-topology: Look for the 32-bit DMIC blob first"

This reverts commit 9ed50a883e37c73a1ecd903cfa9044c048f3326b.

New implementation will incorporate parts of it, but in a differnt
way.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
alsabat failed on mtl-sdw but it looks like a false positive as #4908 have the alsabat fail on the same CI device.

ujfalusi commented:
Changes since v3:
- separate out the revert of #4840 patch
- Update the commit message for the second patch (was the first previously)

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1385,9 +1388,18 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
 		sample_rate = params_rate(params);
+		bit_depth = params_width(params);
+		/*
+		 * Look for 32-bit blob first instead of 16-bit if copier
```
@ujfalusi I feel a bit uncomfortable with clubbing the topology formats with the NHLT blob detection. If the NHLT blobs came from topology always, this would make sense but when the blobs are from the BIOS, why do we need to worry about what the topology contains at all?

ujfalusi replied:
@ranj063, if the copier supports single bit depth (defined by topology) then we can only load an NHLT blob which also supports this and only this bit depth, no?
It does not matter where the NHLT is coming from, the copier config and the blob must be compatible, so we take this restriction from the copier's format support and use it to look for the blob.

I think it is more pragmatic than using 32bit blob unconditionally even if the copier only supports 16bit.

plbossart replied:
@ranj063 you really need to chime in, we can't have two PRs suggested for the same purpose...

plbossart commented:
SOFCI TEST
